include/boost/geometry/algorithms/detail/disjoint/interface.hpp

```diff
@@ -125,104 +131,54 @@ struct disjoint
-
-        template <typename Geometry2>
-        bool operator()(Geometry2 const& geometry2) const
+        bool result = true;
+        geometry::visit([&](auto const& g1, auto const& g2)
         {
-            return disjoint<Geometry1, Geometry2>::apply(m_geometry1, geometry2, m_strategy);
-        }
-    };
-
-    template <typename Strategy>
-    static inline bool apply(Geometry1 const& geometry1,
-                             boost::variant<BOOST_VARIANT_ENUM_PARAMS(T)> const& geometry2,
-                             Strategy const& strategy)
-    {
-        return boost::apply_visitor(visitor<Strategy>(geometry1, strategy), geometry2);
+            result = disjoint
+                <
+                    util::remove_cref_t<decltype(g1)>, util::remove_cref_t<decltype(g2)>
+                >::apply(g1, g2, strategy);
+        }, geometry1, geometry2);
+        return result;
     }
 };
 
-template
-<
-    BOOST_VARIANT_ENUM_PARAMS(typename T1),
-    BOOST_VARIANT_ENUM_PARAMS(typename T2)
->
-struct disjoint
-    <
-        boost::variant<BOOST_VARIANT_ENUM_PARAMS(T1)>,
-        boost::variant<BOOST_VARIANT_ENUM_PARAMS(T2)>
-    >
+// TODO: The complexity is quadratic for two GCs
+//   Decrease e.g. with spatial index,
+template <typename Geometry1, typename Geometry2, bool IsDynamic>
+struct disjoint<Geometry1, Geometry2, IsDynamic, true>
 {
     template <typename Strategy>
-    struct visitor: boost::static_visitor<bool>
-    {
-        Strategy const& m_strategy;
-
-        visitor(Strategy const& strategy)
-            : m_strategy(strategy)
-        {}
-
-        template <typename Geometry1, typename Geometry2>
-        bool operator()(Geometry1 const& geometry1,
-                        Geometry2 const& geometry2) const
-        {
-            return disjoint<Geometry1, Geometry2>::apply(geometry1, geometry2, m_strategy);
-        }
-    };
-
-    template <typename Strategy>
-    static inline bool apply(boost::variant<BOOST_VARIANT_ENUM_PARAMS(T1)> const& geometry1,
-                             boost::variant<BOOST_VARIANT_ENUM_PARAMS(T2)> const& geometry2,
+    static inline bool apply(Geometry1 const& geometry1, Geometry2 const& geometry2,
                              Strategy const& strategy)
     {
-        return boost::apply_visitor(visitor<Strategy>(strategy), geometry1, geometry2);
+        bool result = true;
+        geometry::visit_breadth_first([&](auto const& g1)
+        {
+            geometry::visit_breadth_first([&](auto const& g2)
+            {
+                result = disjoint
+                    <
+                        util::remove_cref_t<decltype(g1)>, util::remove_cref_t<decltype(g2)>
+                    >::apply(g1, g2, strategy);
+                // If any of the combination intersects then the final result is not disjoint
+                //   so break testing.
+                return result;
+            }, geometry2);
+            return result;
+        }, geometry1);
+        return result;
```

Comment:
I don't get the logic yet but I didn't dive into `visit_breadth_first` yet

Reply:
It's based on the assumption that geometry collections can be recursive. So a GC can store GCs. AFAIU this can be the case. So this function traverses the recursive GC in a breadth-first fashion (as opposed to depth-first). And as long as the function/lambda returns `true` the traversal is continued. Maybe I should've called it differently. E.g. `visit_breadth_first_if`?

So here as long as pairs of geometries are keep being disjoint then the traversal should continue. Hence the `result` is returned because it's `true` if pairs of geometries are being disjoint.

Or is it not clear from a different reason?

Reply:
It's clear now. Thank you.